docs/proposals/playbook_consolidation.md

```diff
@@ -0,0 +1,183 @@
+Currently there are two separate sets of initialization playbooks:
+* `playbooks/byo/openshift-cluster/initialize_groups.yml`
+* `playbooks/common/openshift-cluster/std_include.yml`
+
+Although these playbooks are located in the `openshift-cluster` directory they
+are shared by all of the `openshift-*` areas.  These playbooks would be better
+organized in a `playbooks/init` directory collocated with all their related
+playbooks.
+
+In the example below, the following changes have been made:
+* `playbooks/byo/openshift-cluster/initialize_groups.yml` renamed to
+  `playbooks/init/initialize_host_groups.yml`
+* `playbooks/common/openshift-cluster/std_include.yml` renamed to
+  `playbooks/init/main.yml`
+* `- include: playbooks/init/initialize_host_groups.yml` has been added to the
+  top of `playbooks/init/main.yml`
+* All other related files for initialization have been moved to `playbooks/init`
+
+The `initialize_host_groups.yml` playbook is only one play with one task for
+importing variables for inventory group conversions.  This task could be further
+consolidated with the play in `evaluate_groups.yml`.
+
+The new standard initialization playbook would be
+`playbooks/init/main.yml`.
+
+
+```
+ 
+> $ tree openshift-ansible/playbooks/init
+.
+├── evaluate_groups.yml
+├── initialize_facts.yml
+├── initialize_host_groups.yml
+├── initialize_openshift_repos.yml
+├── initialize_openshift_version.yml
+├── main.yml
+├── roles -> ../../roles
+├── validate_hostnames.yml
+└── vars
+    └── cluster_hosts.yml
+```
+
+```yaml
+# openshift-ansible/playbooks/init/main.yml
+---
+- include: initialize_host_groups.yml
+
+- include: evaluate_groups.yml
+
+- include: initialize_facts.yml
+
+- include: validate_hostnames.yml
+
+- include: initialize_openshift_repos.yml
+
+- include: initialize_openshift_version.yml
+```
+
+### `byo` and `common` Playbook Consolidation
+Historically, the `byo` directory coexisted with other platform directories
+which contained playbooks that then called into `common` playbooks to perform
+common installation steps for all platforms.  Since the other platform
+directories have been removed this separation is no longer necessary.
+
+In the example below, the following changes have been made:
+* `playbooks/byo/openshift-master` renamed to
+  `playbooks/openshift-master`
+* `playbooks/common/openshift-master` renamed to
+  `playbooks/openshift-master/private`
+* Original `byo` entry point playbooks have been updated to include their
+  respective playbooks from `private/`.
+* Symbolic links have been updated as necessary
+
+All user consumable playbooks are in the root of `openshift-master` and no entry
+point playbooks exist in the `private` directory.  Maintaining the separation
+between entry point playbooks and the private playbooks allows individual pieces
+of the deployments to be used as needed by other components.
+
+```
+openshift-ansible/playbooks/openshift-master 
+> $ tree
+.
+├── config.yml
+├── private
```

Comment:
Thanks for putting all non-entry points into a separate directory. Just not sure if the `private` is the right name for the folder cause, unfortunately, there is no concept of a private playbook in Ansible :(. Basically anyone installing our solution can include anything. And `there_is_no_entrypoint_playbook` will not help. What about `pages`?

Reply:
It's just a nit, the `private` will do as well.

Reply:
Here is more detail on my thoughts on this organizational approach.  We have talked about where playbooks should go and how to organize them.  Developers are unsure of where to put something and users find playbooks in `byo` or `common` or elsewhere and they are uncertain what is 'runnable' as well.  There was a discussion about renaming `common` to `private` but it seems to make more sense to collapse byo and common into single areas and make a `private` folder for playbooks that should not be directly consumed by users.  We could add a README.md file to the `private` folder indicating such.  Anybody can do anything they want, but we can at least communicate intent of this structure to developers and users.

Reply:
What about playbooks that once we move the initialization bits up into the `init` folder, are only calling into their common component (e.g. `byo/openshift-cluster/openshift-{logging,metrics}.yml`)? Would we collapse that into a single playbook folder and refactor the plays to reflect that?

Reply:
@ewolinetz If I'm understanding your question correctly, does something like this make sense?  The same would apply for any other components we have under `openshift-cluster`.

```
openshift-ansible/playbooks/openshift-metrics
> $ tree
.
├── config.yml        <-- renamed from byo/openshift-cluster/openshift-metrics.yml
└── private
    └── config.yml    <-- renamed from common/openshift-cluster/openshift-metrics.yml
```

Reply:
I think that's what wouldn't make sense to me...
Currently we have `byo/openshift-cluster/openshift-logging.yml`:
```yaml
---
#
# This playbook is a preview of upcoming changes for installing
# Hosted logging on.  See inventory/byo/hosts.*.example for the
# currently supported method.
#
- include: initialize_groups.yml
  tags:
  - always

- include: ../../common/openshift-cluster/std_include.yml
  tags:
  - always

- include: ../../common/openshift-cluster/openshift_logging.yml
```

But the first two includes should be run all the time... unless the thought is we would keep that there so it would be another entry point for just configuring logging/metrics (and then a full OCP install would call into `private/config.yml`)?

Reply:
If a user wanted to only run the logging install, they would use the `playbooks/openshift-logging/config.yml` playbook.
```yaml
# openshift-ansible/playbooks/openshift-logging/config.yml
---
- include: ../init/main.yml

- include: private/config.yml
```
(FYI, I pushed a new commit which references the new `deploy_cluster.yml`.
For the full OCP install, the logging install portion would use `private` from openshift-logging.
```yaml
# openshift-ansible/playbooks/deploy_cluster.yml
---
- include: init/main.yml

- include: openshift-etcd/private/config.yml

- include: openshift-nfs/private/config.yml

- include: openshift-loadbalancer/private/config.yml

- include: openshift-master/private/config.yml

- include: openshift-node/private/config.yml

- include: openshift-glusterfs/private/config.yml

# - include: openshift-hosted/private/config.yml      <-- openshift-hosted is being broken apart now
- include: <how ever we break apart openshift-hosted>

- include: openshift-logging/private/config.yml       <--  Install the logging component

- include: openshift-metrics/private/config.yml       <--  Install the metrics component

- include: openshift-service-catalog/private/config.yml
```